### PR TITLE
concepts: add MEF

### PR DESCRIPTION
**Pull request by rerowep**
Co-Authored-by: Peter Weber <peter.weber@rero.ch>

### CODE REVIEW

File: rero_mef/agents/mef/api.py [hunk 1]
```diff
@@ -48,242 +45,45 @@ class Meta:
...
-        for agent, agent_class in used_classes.items():
-            if verbose:
-                click.echo(f'Get pids from {agent} ...')
-            missing_pids[agent] = {}
-            progress = progressbar(
-                items=agent_class.get_all_pids(),
-                length=agent_class.count(),
-                verbose=verbose
+    def delete_agent(self, agent_record, dbcommit=False, reindex=False):
+        """Delete Agency from record."""
```
Comment (iGormilhit) [inline rero_mef/agents/mef/api.py:75]: Delete agency or agent?

File: rero_mef/agents/viaf/jsonresolvers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/agents/viaf/jsonresolvers/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/agents/viaf/jsonresolvers/viaf_resolver.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/agents/viaf/jsonresolvers/viaf_resolver.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/agents/viaf/jsonresolvers/viaf_resolver.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+
+"""Json resolvers."""
```
Comment (iGormilhit) [inline rero_mef/agents/viaf/jsonresolvers/viaf_resolver.py:19]:
```suggestion
"""JSON resolvers."""
```

File: rero_mef/agents/viaf/serializers.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/agents/viaf/serializers.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/api_mef.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/api_mef.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/api_mef.py [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+    model_cls = None
+    viaf_cls = None
+    search = None
+    mef_type = ''
+
+    @classmethod
+    def get_mef_by_entity_pid(cls, entity_pid, entity_name, pid_only=False):
+        """Get MEF record by entity pid value.
+
+        :param entity_pid: Pid of entety.
```
Comment (iGormilhit) [inline rero_mef/api_mef.py:47]:
```suggestion
        :param entity_pid: Entity PID.
```

File: rero_mef/api_mef.py [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+            else:
+                return cls.get_record_by_pid(mef_pid)
+        except StopIteration:
+            return None
+
+    @classmethod
+    def get_all_mef_pids(cls, entitiy):
+        """Get all MEF pids for entity.
+
+        :param entitiy: Entitiy to search pid for.
```
Comment (iGormilhit) [inline rero_mef/api_mef.py:73]:
```suggestion
        :param entitiy: Entity to search pid for.
```

File: rero_mef/api_mef.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+        :returns: Generator of entitiy pids.
```
Comment (iGormilhit) [inline rero_mef/api_mef.py:74]:
```suggestion
        :returns: Generator of entity pids.
```

File: rero_mef/concepts/cli.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/cli.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/cli.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Click command-line interface for MEF record management."""
```
Comment (iGormilhit) [inline rero_mef/concepts/cli.py:18]:
```suggestion
"""Click command line interface for MEF record management."""
```

File: rero_mef/concepts/mef/__init__.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/api.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/api.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/fetchers.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/fetchers.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/jsonresolvers/__init__.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/jsonresolvers/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/jsonresolvers/mef_resolver.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/jsonresolvers/mef_resolver.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/jsonresolvers/mef_resolver.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+
+"""Json resolvers."""
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/jsonresolvers/mef_resolver.py:19]:
```suggestion
"""JSON resolvers."""
```

File: rero_mef/concepts/mef/jsonschemas/__init__.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/jsonschemas/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/jsonschemas/concepts_mef/mef-concept-v0.0.1.json [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  "type": "object",
+  "required": [
+    "$schema",
+    "pid"
+  ],
+  "additionalProperties": false,
+  "properties": {
+    "$schema": {
+      "title": "Schema",
+      "description": "Schema to MEF concept record.",
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/jsonschemas/concepts_mef/mef-concept-v0.0.1.json:13]:
```suggestion
      "description": "Schema for MEF concept record.",
```

File: rero_mef/concepts/mef/listner.py [hunk 20]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/listner.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/mappings/__init__.py [hunk 21]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/mappings/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/mappings/v7/__init__.py [hunk 22]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/mappings/v7/__init__.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/minters.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/minters.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/minters.py [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+"""Identifier minters."""
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+from .providers import ConceptMefProvider
+
+
+def mef_id_minter(record_uuid, data, provider=ConceptMefProvider,
+                  pid_key='pid', object_type='rec'):
+    """RERIOLS mef conceot id minter."""
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/minters.py:27]:
```suggestion
    """RERO ILS mef concept pid minter."""
```

File: rero_mef/concepts/mef/models.py [hunk 25]
```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/models.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/providers.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/providers.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/providers.py [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+    """Type of persistent identifier."""
+
+    pid_identifier = ConceptMefIdentifier.__tablename__
+    """Identifier for table name"""
+
+    pid_provider = None
+    """Provider name.
+
+    The provider name is not recorded in the PID since the provider does not
+    provide any additional features besides creation of Document ids.
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/providers.py:41]: Is it "Document ids" ? Maybe *Record* ids, or *Concept* ids?

File: rero_mef/concepts/mef/providers.py [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+    """
+
+    default_status = PIDStatus.REGISTERED
+    """Mef IDs are by default registered immediately."""
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/providers.py:45]:
```suggestion
    """Mef IDs are registered immediately by default."""
```

File: rero_mef/concepts/mef/serializers.py [hunk 29]
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/serializers.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/mef/serializers.py [hunk 30]
```diff
@@ -0,0 +1,90 @@
...
+        if request and request.args.get('sources'):
+            sources = []
+            # TODO: add the list of sources into the current_app.config
+            if 'rero' in record:
+                sources.append('rero')
+            record['sources'] = sources
+
+        concept_classes = get_entity_classes()
+        for concept, concept_classe in concept_classes.items():
+            if concept in ['corero']:
```
Comment (iGormilhit) [inline rero_mef/concepts/mef/serializers.py:78]: corero?

File: rero_mef/concepts/serializers.py [hunk 31]
```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/serializers.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/tasks.py [hunk 32]
```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+#
+# RERO MEF
+# Copyright (C) 2020 RERO
```
Comment (iGormilhit) [inline rero_mef/concepts/tasks.py:4]:
```suggestion
# Copyright (C) 2021 RERO
```

File: rero_mef/concepts/tasks.py [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Tasks used by  RERO-MEF."""
```
Comment (iGormilhit) [inline rero_mef/concepts/tasks.py:18]:
```suggestion
"""Tasks used by RERO-MEF."""
```

File: rero_mef/concepts/mef/serializers.py (hunk 30 above)
Comment (rerowep) [reply rero_mef/concepts/mef/serializers.py:78]: It is the pid_type for `Concept RERO`

**Pull request merged**